README.md
# react-native-toast-hook

[![npm version](https://img.shields.io/npm/v/@stratuscode/react-native-toast-hook)](https://www.npmjs.com/package/@stratuscode/react-native-toast-hook)
[![npm downloads](https://img.shields.io/npm/dw/@stratuscode/react-native-toast-hook)](https://www.npmjs.com/package/@stratuscode/react-native-toast-hook)
[![CI Status](https://circleci.com/gh/StratusCode/react-native-toast-hook.svg?style=shield&circle-token=01a30b6f87dedaa5847bf3c85e6b333da11fc64f)](https://app.circleci.com/pipelines/github/StratusCode/react-native-toast-hook)

An animated toast message component for React Native using context API and hooks written in typescript.

## Install

```
npm install --save react-native-toast-hook
```

![ToastSuccess](success-toast.gif)

## Usage

Render the `ToastProvider` component in your app entry file.

```js
// App.jsx
import React from 'react'
import { ToastProvider } from '@stratuscode/react-native-toast-hook'

function App(props) {
	return (
		<>
			<ToastProvider>
				{/* ... */}
			</ToastProvider>
		</>
	)
}

export default App
```

Then use it anywhere in your app with the `useToast` hook, which provides access to `showToast`, `queueToast`, `hideToast`, and `clearToastQueue`:

```js
import React from 'react'
import { useToast } from '@stratuscode/react-native-toast-hook'

function() {
	const { queueToast } = useToast()

	function makeToast() {
		queueToast({
			title: 'Hey you!',
			message: 'The magic medicine works!',
		})
	}

	return (
		<View>
			<Button onPress={makeToast} title="Toast It"/>
		</View>
	)
}
```

If your toasts aren't showing up you might need to manually place the `Toaster`. For example, if you're using `react-navigation` you'll probably want the `Toaster` just before the end of your `</NavigationContainer>`. The `ToastProvider` component will render the `Toaster` component automatically unless you pass `renderToaster={false}` to the `ToastProvider`, so do that and then place your `Toaster` where you want it.

```js
import React from 'react'
import { NavigationContainer } from '@react-navigation/native'
import { createDrawerNavigator } from '@react-navigation/drawer'
import { Toaster } from '@stratuscode/react-native-toast-hook'

const Drawer = createDrawerNavigator()
function Navigation() {
	return (
		<NavigationContainer>
			<Drawer.Navigator {...etc} />
			<Toaster />
		</NavigationContainer>
	)
}
```

## API

### `useToast()`

The useToast hook gives you access to:
- `queueToast: (toast: ToastProps) => void` - shows the toast immediately if one is not currently showing, otherwise queues up the toast to be shown after the current one goes away.
- `showToast: (toast: ToastProps) => void` - cuts in line and shows the toast immediately (the currently shown toast will show again after this important one goes away).
- `hideToast: () => void` - hides the currently shown toast (if any are in queue the next one will show).
- `clearToastQueue: () => void` - hides the currently shown toast and removes the entire queue.

See the [types.ts](https://github.com/StratusCode/react-native-toast-hook/blob/master/src/types/index.ts) file for the most up-to-date options in `ToastProps` and for more advanced usages, but the gist is:

```ts
queueToast({
	title?: 'Top line, bolded text',
	message?: 'Bottom line, normal text',
	type?: 'info' | 'success' | 'warning' | 'error',
	position?: 'top' | 'bottom',
	visibilityTime?: 4000,
	autoHide?: true,
	onPress?: (toast: ToastProps) => void  // callback for when the toast is pressed, e.g. to handle deep navigation
})
```

## Global customizations

`ToastProvider` accepts some props to set global defaults. 
- The `defaults` prop takes the same object that `queueToast`/`showToast` take and will cause those functions to fallback on the default values for each prop that their object doesn't have defined.
- The `colors` prop lets you override any colors used (see `ToastColors` in [types.ts](https://github.com/StratusCode/react-native-toast-hook/blob/master/src/types/index.ts))
- The `customToasts` lets you configure your own toast components from scratch or override the default ones (see below).

## Customizing the toast types

If you want to add custom types, or overwrite the existing ones, you can add a `customToasts` prop when rendering the `ToastProvider`. You can import the components that are used by default and customize them, use BaseToast and its render functions for almost total control, or provide your own toast implementation for total control.

```js
// App.jsx
import React from 'react'
import { Text } from 'react-native'
import { ToastProvider, BaseToast, InfoToast, WarningToast } from '@stratuscode/react-native-toast-hook'
import Icon from 'react-native-vector-icons/FontAwesome';

const toastConfig = {
	// modify only the formatting of the default info toast's title
	info: (toast) => (
		<InfoToast renderTitle={(toast) => (
			<Text style={{ fontWeight: 'normal', color: 'fuchsia' }}>
				{toast.title}
			</Text>
		)} />
	),
	// override existing success toast with entirely custom toast
	success: (toast) => (
		<View style={{ height: 60, width: '100%', backgroundColor: 'pink' }}>
			<Text>{toast.title}</Text>
			<Text>{toast.message}</Text>
		</View>
	),
	// modify only the icon of the default warning toast
	warning: (toast) => (
		<WarningToast renderIcon={({color}) => (
			<Icon name="rocket" size={24} color={color} />
		)} />
	),
	// create a new type of toast, just supply 'any_custom_type' as the `type` prop to `queueToast`/`showToast`
	any_custom_type: (toast) => {},
}

function App(props) {
	return (
		<>
			<ToastProvider customToasts={toastConfig}>
				{/* ... */}
			</ToastProvider>
		</>
	)
}

export default App
```

Then just use the library as before

```js
queueToast({
	type: 'any_custom_type',
	title: 'Woohoo',
	message: 'My own custom type (:',
})
```

## Credits

- This library was originally forked from Calin Tamas' [react-native-toast-message](https://github.com/calintamas/react-native-toast-message).
- Icons (`info`, `success`, `warning`, `error`, and `close`) came from [Pixel perfect](https://www.flaticon.com/authors/pixel-perfect)'s [Basic UI Icon Pack](https://www.flaticon.com/packs/basic-ui-4) from [flaticon.com](https://www.flaticon.com).


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

⚠️ The changelog should be **human-readable**, so everything that's added here should be easy to understand without additional lookups and checks

Headers are one of:

-   `Added`, `Changed`, `Removed` or `Fixed`.

## [2.0.5]

### Fixed

-   Fixed autohide

### Changed

-   Changed topOffset's default to 40 due to iPhone notches

## [2.0.4]

### Changed

-   Forked from v1.3.4 of Calin Tamas' [react-native-toast-message](https://github.com/calintamas/react-native-toast-message)
-   Rewrote component in typescript using react context and hooks
-   Switched out raster images for SVGs

### Added

-   Added react and react-native as peer dependencies
-   Added new peer dependency [react-native-svg](https://www.npmjs.com/package/react-native-svg) for SVG icons
-   Added ToastProvider and useToast as the new API
-   Added optional onPress to toast
-   Added a queue for toasts -- if you queue a toast while one is already shown it will appear after the current one is gone

## [2.0.(0-3)]

-   Broken builds. Removed from npm. Use 2.0.4+


src/types/index.ts
import { ColorValue } from "react-native"
import { Color as SvgColor } from "react-native-svg"

export interface ToastProps {
	/**
	 * Use predefined types for standard colors/icons, or configure your own types on the provider and use them here. Defaults to 'info'
	 */
	type?: "info" | "success" | "error" | "warning" | string // TODO: type-safe the custom config values with generics. (this is hard)
	/**
	 * Where to show the toasts. Defaults to 'top'
	 */
	position?: "top" | "bottom"
	/**
	 * Bolded top line of text
	 */
	title?: string
	/**
	 * Bottom line of text
	 */
	message?: string
	/**
	 * True if the message hide automatically, false if the user should have to interact with it. Defaults to true
	 */
	autoHide?: boolean
	/**
	 * Milliseconds before autoHide happens. Defaults to 4000
	 */
	visibilityTime?: number
	/**
	 * Height of the toast component. Defaults to 60
	 */
	height?: number
	/**
	 * Points from the top of the screen (when position='top'). Defaults to 40
	 */
	topOffset?: number
	/**
	 * Points from the bottom of the screen (when position='bottom'). Defaults to 40
	 */
	bottomOffset?: number
	/**
	 * friction value for the show/hide animation. Defaults to 8
	 */
	friction?: number
	/**
	 * Callback when a toast is shown
	 */
	onShow?: (toast: ToastProps) => void
	/**
	 * Callback when a toast is queued
	 */
	onQueue?: (toast: ToastProps) => void
	/**
	 * Callback when a toast is hidden (automatically or by the user)
	 */
	onHide?: (toast: ToastProps) => void
	/**
	 * Callback when a toast is pressed (prevents onHide from firing)
	 */
	onPress?: (toast: ToastProps) => void
}

export interface ToastProviderProps {
	/**
	 * Optionally provide defaults for all toasts (settings here can be overridden when showing or queueing a toast)
	 */
	defaults?: ToastProps
	/**
	 * Optionally override default colors
	 */
	colors?: ToastColors
	/**
	 * Configure your own toast components to render. Object keys maps to the `ToastProps` `type` prop while the values are render functions to render the toast.
	 */
	customToasts?: ToastComponentsConfig
	/**
	 * The Toaster will be rendered for you by default unless you set this to false.
	 * If you need to render it in a place other than where the provider is placed the useToast hook will give you the Toaster you need to render.
	 * react-navigation is known to interfere with the rendering of the toasts. Place the Toaster just before your </NavigationContainer> to fix it.
	 */
	renderToaster?: boolean
}

export interface UseToastHook {
	/**
	 * Pops up a toast (jumps to the front of the queue and hides the existing toast if there is one)
	 */
	showToast: (props: ToastProps) => void
	/**
	 * Queues up a toast (or shows it instantly if there are no toasts showing and in the queue)
	 */
	queueToast: (props: ToastProps) => void
	/**
	 * Hides the current toast, which will show the next toast in the queue if there is one
	 */
	hideToast: () => void
	/**
	 * Purges the toast queue (but does not hide the one being shown)
	 */
	clearToastQueue: () => void
}

export interface IToastContext extends ToastContextSettings, UseToastHook {
	toasts: Readonly<ToastProps[]>
	activeToast?: ToastProps | null
}

type OptionalSettings = Pick<
	ToastProps,
	"title" | "message" | "onShow" | "onQueue" | "onHide" | "onPress"
>

interface ToastContextSettings {
	colors: Required<ToastColors>
	customToasts?: ToastComponentsConfig
	defaults: Required<Omit<ToastProps, keyof OptionalSettings>> & OptionalSettings
}

export type Color = ColorValue & SvgColor
export interface ToastColors {
	/**
	 * Defaults to #fff
	 */
	background?: Color
	/**
	 * Defaults to #D8D8D8
	 */
	leftBorder?: Color
	/**
	 * Defaults to #221D23
	 */
	title?: Color
	/**
	 * Defaults to #221D23
	 */
	message?: Color
	/**
	 * Defaults to #221D23
	 */
	closeIcon?: Color
	/**
	 * Accent color for the info toast type. Defaults to #00c9ff
	 */
	info?: Color
	/**
	 * Accent color for the error toast type. Defaults to #fe3618
	 */
	error?: Color
	/**
	 * Accent color for the warning toast type. Defaults to #fcc603
	 */
	warning?: Color
	/**
	 * Accent color for the success toast type. Defaults to #00ac31
	 */
	success?: Color
}

export interface ToastComponentsConfig {
	[x: string]: (toast: BaseToastProps) => React.ReactElement
}

export interface SvgProps {
	color?: Color
	size?: number
}

export interface BaseToastProps extends ToastProps {
	/**
	 * Color of the left accent and icon
	 */
	color?: Color
	/**
	 * Icon to show for the toast. This library does not include any of the icon libraries but has SVG defaults for info, success, warning, error, and close.
	 * You can use an SVG component like this library does or provide an icon from any library you wish (make sure to size and color it properly).
	 */
	iconElement?: React.ReactNode
	/**
	 * The close button should fire this to hide the toast.
	 */
	onClose: () => void
	/**
	 * Optional render function to render the icon (and its wrapper).
	 */
	renderIcon?: (toast: BaseToastProps) => React.ReactNode
	/**
	 * Optional render function to render the title.
	 */
	renderTitle?: (toast: BaseToastProps) => React.ReactNode
	/**
	 * Optional render function to render the message.
	 */
	renderMessage?: (toast: BaseToastProps) => React.ReactNode
	/**
	 * Optional render function to render the close button.
	 */
	renderCloseButton?: (toast: BaseToastProps) => React.ReactNode
}

export type Toaster = React.FC
